File: frontend/src/api/services/SprintService.ts
import apiClient from "@api/client";
import type { Sprint } from "@/types";

const parseSprintDates = (sprint: any): Sprint => ({
  ...sprint,
  startDate: new Date(sprint.startDate),
  finishDate: new Date(sprint.finishDate),
});

export const sprintService = {
  // --- PUT --- //
  // Обновление спринта
  updateSprint: async (sprintData: Sprint): Promise<Sprint> => {
    const response = await apiClient.put("/sprint/update", sprintData);
    return parseSprintDates(response.data);
  },

  // --- POST --- //
  // Создание спринта
  createSprint: async (sprintData: Omit<Sprint, "id">): Promise<Sprint> => {
    const response = await apiClient.post("/sprint/create", sprintData);
    return parseSprintDates(response.data);
  },

  // --- PATCH --- //
  // Старт спринта
  startSprint: (sprintId: string): Promise<void> => {
    return apiClient.patch(`/sprint/start/${sprintId}`);
  },

  // Завершение спринта
  completeSprint: (sprintId: string, report: string): Promise<void> => {
    return apiClient.patch(`/sprint/complete/${sprintId}`, report, {
      headers: { "Content-Type": "text/plain" },
    });
  },

  // --- GET --- //
  // Получение всех спринтов скрама
  getAllSprints: async (scrumId: string): Promise<Sprint[]> => {
    const response = await apiClient.get(`/sprint/all/${scrumId}`);
    return response.data.map(parseSprintDates);
  },

  // Получение активных спринтов
  getActiveSprints: async (scrumId: string): Promise<Sprint[]> => {
    const response = await apiClient.get(`/sprint/active/${scrumId}`);
    return response.data.map(parseSprintDates);
  },

  // --- DELETE --- //
  // Удаление спринта
  deleteSprint: (sprintId: string): Promise<void> => {
    return apiClient.delete(`/sprint/delete/${sprintId}`);
  },
};

File: backend/src/main/java/com/tyuiu/backend/services/SprintService.java
package com.tyuiu.backend.services;

import com.tyuiu.backend.models.dto.SprintDTO;
import com.tyuiu.backend.models.entities.Sprint;
import com.tyuiu.backend.models.enums.SprintStatus;
import com.tyuiu.backend.models.mappers.SprintMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.data.r2dbc.core.R2dbcEntityTemplate;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import static org.springframework.data.relational.core.query.Criteria.where;
import static org.springframework.data.relational.core.query.Query.query;
import static org.springframework.data.relational.core.query.Update.update;

@Service
@RequiredArgsConstructor
public class SprintService {
    private final R2dbcEntityTemplate template;
    private final SprintMapper sprintMapper;

    public Flux<SprintDTO> getAllSprints(String scrumId) {
        return template.select(query(where("scrum_id").is(scrumId)), Sprint.class)
                .map(sprintMapper::toDTO);
    }

    public Flux<SprintDTO> getActiveSprints(String scrumId) {
        return template.select(query(where("scrum_id").is(scrumId)
                .and(where("status").is(SprintStatus.ACTIVE))), Sprint.class)
                .map(sprintMapper::toDTO);
    }

    public Mono<SprintDTO> createSprint(SprintDTO sprintDTO) {
        return template.insert(sprintMapper.toEntity(sprintDTO))
                .map(sprintMapper::toDTO);
    }

    public Mono<SprintDTO> updateSprint(SprintDTO sprintDTO) {
        return template.update(sprintMapper.toEntity(sprintDTO))
                .flatMap(updated -> template.selectOne(query(where("id").is(sprintDTO.getId())), Sprint.class))
                .map(sprintMapper::toDTO);
    }

    public Mono<Void> startSprint(String sprintId) {
        return template.update(query(where("id").is(sprintId).and(where("status").is(SprintStatus.ACTIVE))),
                        update("status", SprintStatus.DONE), Sprint.class)
                .then(template.update(query(where("id").is(sprintId)),
                update("status", SprintStatus.ACTIVE), Sprint.class)).then();
    }

    public Mono<Void> completeSprint(String sprintId, String report) {
        return template.update(query(where("id").is(sprintId)),
                update("status", SprintStatus.DONE)
                        .set("report", report), Sprint.class)
                .then();
    }

    public Mono<Void> deleteSprint(String sprintId) {
        return template.delete(query(where("id").is(sprintId)), Sprint.class).then();
    }
}
